kernel/device.c

```diff
@@ -9,89 +9,93 @@
-extern struct device __device_POST_KERNEL_start[];
-extern struct device __device_APPLICATION_start[];
-extern struct device __device_init_end[];
+extern struct init_entry __init_start[];
+extern struct init_entry __init_PRE_KERNEL_1_start[];
+extern struct init_entry __init_PRE_KERNEL_2_start[];
+extern struct init_entry __init_POST_KERNEL_start[];
+extern struct init_entry __init_APPLICATION_start[];
+extern struct init_entry __init_end[];
 
 #ifdef CONFIG_SMP
-extern struct device __device_SMP_start[];
+extern struct device __init_SMP_start[];
 #endif
 
+extern struct device __device_start[];
+extern struct device __device_end[];
+
 #ifdef CONFIG_DEVICE_POWER_MANAGEMENT
 extern u32_t __device_busy_start[];
 extern u32_t __device_busy_end[];
 #define DEVICE_BUSY_SIZE (__device_busy_end - __device_busy_start)
 #endif
 
 /**
- * @brief Execute all the device initialization functions at a given level
+ * @brief Execute all the init entry initialization functions at a given level
  *
- * @details Invokes the initialization routine for each device object
- * created by the DEVICE_INIT() macro using the specified level.
- * The linker script places the device objects in memory in the order
+ * @details Invokes the initialization routine for each init entry object
+ * created by the INIT_ENTRY_DEFINE() macro using the specified level.
+ * The linker script places the init entry objects in memory in the order
  * they need to be invoked, with symbols indicating where one level leaves
  * off and the next one begins.
  *
  * @param level init level to run.
  */
-void z_sys_device_do_config_level(s32_t level)
+void z_sys_init_run_level(s32_t level)
 {
-	struct device *info;
-	static struct device *config_levels[] = {
-		__device_PRE_KERNEL_1_start,
-		__device_PRE_KERNEL_2_start,
-		__device_POST_KERNEL_start,
-		__device_APPLICATION_start,
+	static struct init_entry *levels[] = {
+		__init_PRE_KERNEL_1_start,
+		__init_PRE_KERNEL_2_start,
+		__init_POST_KERNEL_start,
+		__init_APPLICATION_start,
 #ifdef CONFIG_SMP
-		__device_SMP_start,
+		__init_SMP_start,
 #endif
 		/* End marker */
-		__device_init_end,
+		__init_end,
 	};
+	struct init_entry *entry;
 
-	for (info = config_levels[level]; info < config_levels[level+1];
-								info++) {
+	for (entry = levels[level]; entry < levels[level+1]; entry++) {
+		struct device *dev = entry->dev;
 		int retval;
-		const struct device_config *device_conf = info->config;
 
-		retval = device_conf->init(info);
+		if (dev != NULL) {
+			z_object_init(dev);
+		}
+
+		retval = entry->init(dev);
 		if (retval != 0) {
-			/* Initialization failed. Clear the API struct so that
-			 * device_get_binding() will not succeed for it.
-			 */
-			info->driver_api = NULL;
-		} else {
-			z_object_init(info);
+			if (dev) {
+				/* Initialization failed. Clear the API struct
+				 * so that device_get_binding() will not succeed
+				 * for it.
+				 */
+				dev->driver_api = NULL;
```

Comment:
because of that dev must be in RAM. Any suggestions how to avoid that?

Reply:
Read #22941 carefully (specially points 7, 4 and 5). 

It's already being tackled.